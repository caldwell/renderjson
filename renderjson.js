// Copyright © 2013-2014 David Caldwell <david@porkrind.org>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

// Usage
// -----
// The module exports one entry point, the `renderjson()` function. It takes in
// the JSON you want to render as a single argument and returns an HTML
// element.
//
// Options
// -------
// renderjson.set_icons("+", "-")
//   This Allows you to override the disclosure icons.
//
// renderjson.set_show_to_level(level)
//   Pass the number of levels to expand when rendering. The default is 0, which
//   starts with everything collapsed. As a special case, if level is the string
//   "all" then it will start with everything expanded.
//
// renderjson.set_expandable_string_length(length)
//   String values will become expandable if they are longer than passed length.
//   The default is 100. As a special case, if length is the string "none" then no
//   string will become expandable.
//
// renderjson.set_sort_objects(sort_bool)
//   Sort objects by key (default: false)
//
// Theming
// -------
// The HTML output uses a number of classes so that you can theme it the way
// you'd like:
//     .disclosure    ("⊕", "⊖")
//     .syntax        (",", ":", "{", "}", "[", "]")
//     .string        (includes quotes)
//     .number
//     .boolean
//     .key           (object key)
//     .keyword       ("null", "undefined")
//     .object.syntax ("{", "}")
//     .array.syntax  ("[", "]")

var module;
(module||{}).exports = renderjson = (function() {
    var themetext = function(/* [class, text]+ */) {
        var spans = [];
        while (arguments.length)
            spans.push(append(span(Array.prototype.shift.call(arguments)),
                              text(Array.prototype.shift.call(arguments))));
        return spans;
    };
    var append = function(/* el, ... */) {
        var el = Array.prototype.shift.call(arguments);
        for (var a=0; a<arguments.length; a++)
            if (arguments[a].constructor == Array)
                append.apply(this, [el].concat(arguments[a]));
            else
                el.appendChild(arguments[a]);
        return el;
    };
    var prepend = function(el, child) {
        el.insertBefore(child, el.firstChild);
        return el;
    }
    var isempty = function(obj) { for (var k in obj) if (obj.hasOwnProperty(k)) return false;
                                  return true; }
    var text = function(txt) { return document.createTextNode(txt) };
    var div = function() { return document.createElement("div") };
    var span = function(classname) { var s = document.createElement("span");
                                     if (classname) s.className = classname;
                                     return s; };
    var A = function A(txt, classname, callback) { var a = document.createElement("a");
                                                   if (classname) a.className = classname;
                                                   a.appendChild(text(txt));
                                                   a.href = '#';
                                                   a.onclick = function() { callback(); return false; };
                                                   return a; };

    function _renderjson(json, indent, dont_indent, show_level, expandable_length, sort_objects) {
        var my_indent = dont_indent ? "" : indent;

        var disclosure = function(open, close, type, builder) {
            var content;
            var empty = span(type);
            var show = function() { if (!content) append(empty.parentNode,
                                                         content = prepend(builder(),
                                                                           A(renderjson.hide, "disclosure",
                                                                             function() { content.style.display="none";
                                                                                          empty.style.display="inline"; } )));
                                    content.style.display="inline";
                                    empty.style.display="none"; };
            append(empty,
                   A(renderjson.show, "disclosure", show),
                   themetext(type+ " syntax", open),
                   A(" ... ", null, show),
                   themetext(type+ " syntax", close));

            var el = append(span(), text(my_indent.slice(0,-1)), empty);
            if (show_level > 0)
                show();
            return el;
        };

        if (json === null) return themetext(null, my_indent, "keyword", "null");
        if (json === void 0) return themetext(null, my_indent, "keyword", "undefined");
        // Strings, numbers and bools.
        if (typeof(json) != "object") {
            var value = JSON.stringify(json);
            if (value.charAt(0) != '"' || value.length <= expandable_length)
                return themetext(null, my_indent, typeof(json), value);

            return disclosure('"', '"', "string", function () {
                return append(span("string"), themetext(null, my_indent, "string", value));
            });
        }

        if (json.constructor == Array) {
            if (json.length == 0) return themetext(null, my_indent, "array syntax", "[]");

            return disclosure("[", "]", "array", function () {
                var as = append(span("array"), themetext("array syntax", "[", null, "\n"));
                for (var i=0; i<json.length; i++)
                    append(as,
                           _renderjson(json[i], indent+"    ", false, show_level-1, expandable_length, sort_objects),
                           i != json.length-1 ? themetext("syntax", ",") : [],
                           text("\n"));
                append(as, themetext(null, indent, "array syntax", "]"));
                return as;
            });
        }

        // object
        if (isempty(json))
            return themetext(null, my_indent, "object syntax", "{}");

        return disclosure("{", "}", "object", function () {
            var os = append(span("object"), themetext("object syntax", "{", null, "\n"));
            for (var k in json) var last = k;
            var keys = Object.keys(json);
            if (sort_objects)
                keys = keys.sort();
            for (var i in keys) {
                var k = keys[i];
                append(os, themetext(null, indent+"    ", "key", '"'+k+'"', "object syntax", ': '),
                       _renderjson(json[k], indent+"    ", true, show_level-1, expandable_length, sort_objects),
                       k != last ? themetext("syntax", ",") : [],
                       text("\n"));
            }
            append(os, themetext(null, indent, "object syntax", "}"));
            return os;
        });
    }

    var renderjson = function renderjson(json)
    {
        var pre = append(document.createElement("pre"), _renderjson(json, "", false, renderjson.show_to_level, renderjson.expandable_string_length, renderjson.sort_objects));
        pre.className = "renderjson";
        return pre;
    };
    // 100 characters is the default expandable string length.
    renderjson.expandable_string_length = 100;
    renderjson.set_icons = function(show, hide) { renderjson.show = show;
                                                  renderjson.hide = hide;
                                                  return renderjson; };
    renderjson.set_show_to_level = function(level) { renderjson.show_to_level = typeof level == "string" &&
                                                                                level.toLowerCase() === "all" ? Number.MAX_VALUE
                                                                                                              : level;
                                                     return renderjson; };
    renderjson.set_expandable_string_length = function(length) { renderjson.expandable_string_length = typeof length == "string" &&
                                                                 length.toLowerCase() === "none" ? Number.MAX_VALUE : length;
                                                                 return renderjson; };
    renderjson.set_sort_objects = function(sort_bool) { renderjson.sort_objects = sort_bool;
                                                        return renderjson; };
    // Backwards compatiblity. Use set_show_to_level() for new code.
    renderjson.set_show_by_default = function(show) { renderjson.show_to_level = show ? Number.MAX_VALUE : 0;
                                                      return renderjson; };
    renderjson.set_icons('⊕', '⊖');
    renderjson.set_show_by_default(false);
    renderjson.set_sort_objects(false);
    return renderjson;
})();
