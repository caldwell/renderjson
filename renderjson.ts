// Copyright © 2013-2017 David Caldwell <david@porkrind.org>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

// Usage
// -----
// The module exports one entry point, the `renderjson()` function. It takes in
// the JSON you want to render as a single argument and returns an HTML
// element.
//
// Options
// -------
// see function documentation
//
// Theming
// -------
// The HTML output uses a number of classes so that you can theme it the way
// you'd like:
//     .disclosure    ("⊕", "⊖")
//     .syntax        (",", ":", "{", "}", "[", "]")
//     .string        (includes quotes)
//     .number
//     .boolean
//     .key           (object key)
//     .keyword       ("null", "undefined")
//     .object.syntax ("{", "}")
//     .array.syntax  ("[", "]")

export interface Options{
    /** show icon */
    show:string
    /** hide icon */
    hide:string
    replacer?: (key: string, value: any) => any
    show_to_level: number
    sort_objects: boolean
    collapse_msg:Function
    property_list:any[]
    max_string_length:number
}

export {} // hack to get typescript to stop complaining about window being redeclared

var module, window, define:Function, renderjson=(function() {
    const themetext = function(...args: string[] /* [class, text]+ */) {
        const spans = [];
        while (args.length)
            spans.push(append(span(Array.prototype.shift.call(args)),
                              text(Array.prototype.shift.call(args))));
        return spans;
    };
    const append = function(...args:(Node|null|Node[])[] /* el, ... */) {
        const el:Node|null = Array.prototype.shift.call(args);
        for (let a=0; a<args.length; a++)
            if (args[a].constructor == Array)
                append.apply(this, [el].concat(args[a]));
            else
                el.appendChild(<Node>args[a]);
        return el;
    };
    const prepend = function<T extends Node>(el:Node, child:T) {
        el.insertBefore<T>(child, el.firstChild);
        return el;
    }
    const isempty = function(obj:{}, pl:any[]) { const keys = pl || Object.keys(obj);
                                      for (const i in keys) if (Object.hasOwnProperty.call(obj, keys[i])) return false;
                                      return true; }
    const text = function(txt:string) { return document.createTextNode(txt) };
    const span = function(classname?:string) { const s = document.createElement("span");
                                     if (classname) s.className = classname;
                                     return s; };
    const A = function A(txt:string, classname:string, callback:Function) { const a = document.createElement("a");
                                                   if (classname) a.className = classname;
                                                   a.appendChild(text(txt));
                                                   a.href = '#';
                                                   a.onclick = function(e) { callback(); if (e) e.stopPropagation(); return false; };
                                                   return a; };

    function _renderjson(json:any, indent:string, dont_indent:boolean, show_level:number, options:Options) {
        const my_indent = dont_indent ? "" : indent;

        const disclosure = function(open:string, placeholder:string, close:string, type:string, builder:Function) {
            let content:HTMLElement;
            const empty = span(type);
            const show = function() { if (!content) append(empty.parentNode,
                                                         content = <HTMLElement>prepend(builder(),
                                                                           A(options.hide, "disclosure",
                                                                             function() { content.style.display="none";
                                                                                          empty.style.display="inline"; } )));
                                    content.style.display="inline";
                                    empty.style.display="none"; };
            append(empty,
                   A(options.show, "disclosure", show),
                   themetext(type+ " syntax", open),
                   A(placeholder, null, show),
                   themetext(type+ " syntax", close));

            const el = append(span(), text(my_indent.slice(0,-1)), empty);
            if (show_level > 0 && type != "string")
                show();
            return el;
        };

        if (json === null) return themetext(null, my_indent, "keyword", "null");
        if (json === void 0) return themetext(null, my_indent, "keyword", "undefined");

        if (typeof(json) == "string" && json.length > options.max_string_length)
            return disclosure('"', json.substr(0,options.max_string_length)+" ...", '"', "string", function () {
                return append(span("string"), themetext(null, my_indent, "string", JSON.stringify(json)));
            });

        if (typeof(json) != "object" || [Number, String, Boolean, Date].indexOf(json.constructor) >= 0) // Strings, numbers and bools
            return themetext(null, my_indent, typeof(json), JSON.stringify(json));

        if (json.constructor == Array) {
            if (json.length == 0) return themetext(null, my_indent, "array syntax", "[]");

            return disclosure("[", options.collapse_msg((<any[]>json).length), "]", "array", function () {
                const as = append(span("array"), themetext("array syntax", "[", null, "\n"));
                for (let i=0; i<(<any[]>json).length; i++)
                    append(as,
                           _renderjson(options.replacer.call(json, i, (<any[]>json)[i]), indent+"    ", false, show_level-1, options),
                           i != json.length-1 ? themetext("syntax", ",") : [],
                           text("\n"));
                append(as, themetext(null, indent, "array syntax", "]"));
                return as;
            });
        }

        // object
        if (isempty(json, options.property_list))
            return themetext(null, my_indent, "object syntax", "{}");

        return disclosure("{", options.collapse_msg(Object.keys(json).length), "}", "object", function () {
            const os = append(span("object"), themetext("object syntax", "{", null, "\n"));
            for (const k in json) var last = k;
            let keys = options.property_list || Object.keys(json);
            if (options.sort_objects)
                keys = keys.sort();
            for (const i in keys) {
                const k = keys[i];
                if (!(k in json)) continue;
                append(os, themetext(null, indent+"    ", "key", '"'+k+'"', "object syntax", ': '),
                       _renderjson(options.replacer.call(json, k, json[k]), indent+"    ", true, show_level-1, options),
                       k != last ? themetext("syntax", ",") : [],
                       text("\n"));
            }
            append(os, themetext(null, indent, "object syntax", "}"));
            return os;
        });
    }

    const renderjson = function renderjson(json:any)
    {
        //@ts-ignore
        const options:Options = new Object(renderjson.options);
        options.replacer = typeof(options.replacer) == "function" ? options.replacer : function<T>(k:any,v:T) { return v; };
        const pre = <Element>append(document.createElement("pre"), _renderjson(json, "", false, options.show_to_level, options));
        pre.className = "renderjson";
        return pre;
    }

    /**
     * This allows you to override the disclosure icons.
     */
    renderjson.set_icons = function(show:string, hide:string) { 
        renderjson.options.show = show;
        renderjson.options.hide = hide;
        return renderjson;
    };

    /**
     * Pass the number of levels to expand when rendering. The default is 0, which
     * starts with everything collapsed. As a special case, if level is the string
     * "all" then it will start with everything expanded.
     */
    renderjson.set_show_to_level = function(level:string|number) { 
        renderjson.options.show_to_level = typeof level == "string" &&
            level.toLowerCase() === "all" ? Number.MAX_VALUE : <number>level;
        return renderjson;
    };

    /**
     * Strings will be truncated and made expandable if they are longer than
     * `length`. As a special case, if `length` is the string "none" then
     * there will be no truncation. The default is "none".
     */
    renderjson.set_max_string_length = function(length:string|number) {
        renderjson.options.max_string_length = typeof length == "string" &&
             length.toLowerCase() === "none" ? Number.MAX_VALUE : <number>length;
        return renderjson;
    };

    /**
     * Sort objects by key (default: false)
     */
    renderjson.set_sort_objects = function(sort:boolean) {
        renderjson.options.sort_objects = sort;
        return renderjson;
    };

    /**
     * Equivalent of JSON.stringify() `replacer` argument when it's a function
     */
    renderjson.set_replacer = function(replacer?:(key: string, value: any) => any) {
        renderjson.options.replacer = replacer;
        return renderjson;
    };

    /**
     * Accepts a function (len:number):string => {} where len is the length of the
     * object collapsed.  Function should return the message displayed when a
     * object is collapsed.  The default message is "X items"
     */
    renderjson.set_collapse_msg = function(collapse_msg:(len:number)=>string) {
        renderjson.options.collapse_msg = collapse_msg;
        return renderjson;
    };

    /**
     * Equivalent of JSON.stringify() `replacer` argument when it's an array
     */
    renderjson.set_property_list = function(prop_list?:any[]) {
        renderjson.options.property_list = prop_list;
        return renderjson;
    };
    // Backwards compatiblity. Use set_show_to_level() for new code.
    renderjson.set_show_by_default = function(show:boolean) {
        renderjson.options.show_to_level = show ? Number.MAX_VALUE : 0;
        return renderjson;
    };
    renderjson.options = <Options>{};
    renderjson.set_icons('⊕', '⊖');
    renderjson.set_show_by_default(false);
    renderjson.set_sort_objects(false);
    renderjson.set_max_string_length("none");
    renderjson.set_replacer(void 0);
    renderjson.set_property_list(void 0);
    renderjson.set_collapse_msg(function(len:number) { return len + " item" + (len==1 ? "" : "s") })
    return renderjson;
})();

// https://requirejs.org/docs/api.html#define
// define in order to support AMD
if (define) define({renderjson:renderjson})
// @ts-ignore
else (module||{}).exports = (window||{}).renderjson = renderjson;

export default renderjson;