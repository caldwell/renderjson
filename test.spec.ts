import renderjson from './renderjson'

test('has appropriate html', ()=>{
    const e = renderjson({})
    expect(e.className).toBe("renderjson")
    expect(e.innerHTML).toBe('<span></span><span class=\"object syntax\">{}</span>')
})

test('handles empty json', ()=>{
    const e = renderjson({})
    expect(e.textContent).toBe("{}")
})

test('handles string', ()=>{
    const e = renderjson("hello world")
    console.log(e)
    expect(e.textContent).toBe('"hello world"')
})

test('arrays are truncated', ()=>{
    const e = renderjson([1,2,3])
    console.log(e)
    expect(e.textContent).toBe("⊕[3 items]")
})

test('json is truncated', ()=>{
    const e = renderjson({"a":3})
    console.log(e)
    expect(e.textContent).toBe('⊕{1 item}')
})

test('icons can be set', ()=>{
    const e = renderjson.set_icons("+", "-")([1,2,3])
    expect(e.textContent).toBe("+[3 items]")
    // reset to default
    renderjson.set_icons("⊕","⊖")
})

test('show to level', ()=>{
    const e = renderjson.set_show_to_level(1)([1,2,3])
    expect(e.firstChild.lastChild.textContent).toBe(`⊖[
    1,
    2,
    3
]`)
    renderjson.set_show_to_level(0)
})

test('max string length is none by default', ()=>{
    // personally I think the max string length should be set by default
    // but I can leave that for later, right now I just want to write tests
    const e = renderjson("a".repeat(10000))
    console.log(e)
    expect(e.textContent).toBe('"' + "a".repeat(10000) + '"')
})

test('has appropriate html for complicated json', ()=>{
    // we want to verify entire json
    renderjson.set_show_to_level(999)

    const e = renderjson({
        "glossary": {
            "title": "example glossary",
            "comprehensive": true,
            "link": undefined,
            "count": 1,
            "GlossDiv": {
                "title": "S",
                "GlossList": {
                    "GlossEntry": {
                        "ID": "SGML",
                        "SortAs": "SGML",
                        "GlossTerm": "Standard Generalized Markup Language",
                        "Acronym": "SGML",
                        "Abbrev": "ISO 8879:1986",
                        "GlossDef": {
                            "para": "A meta-markup language, used to create markup languages such as DocBook.",
                            "GlossSeeAlso": ["GML", "XML"]
                        },
                        "GlossSee": "markup"
                    }
                }
            }
        }
    })
    expect(e.innerHTML).toMatchSnapshot()
    
    renderjson.set_show_to_level(0)
})