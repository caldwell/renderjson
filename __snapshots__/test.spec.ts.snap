// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has appropriate html for complicated json 1`] = `
"<span><span class=\\"object\\" style=\\"display: none;\\"><a class=\\"disclosure\\" href=\\"#\\">⊕</a><span class=\\"object syntax\\">{</span><a href=\\"#\\">1 item</a><span class=\\"object syntax\\">}</span></span><span class=\\"object\\" style=\\"display: inline;\\"><a class=\\"disclosure\\" href=\\"#\\">⊖</a><span class=\\"object syntax\\">{</span><span>
</span><span>    </span><span class=\\"key\\">\\"glossary\\"</span><span class=\\"object syntax\\">: </span><span><span class=\\"object\\" style=\\"display: none;\\"><a class=\\"disclosure\\" href=\\"#\\">⊕</a><span class=\\"object syntax\\">{</span><a href=\\"#\\">5 items</a><span class=\\"object syntax\\">}</span></span><span class=\\"object\\" style=\\"display: inline;\\"><a class=\\"disclosure\\" href=\\"#\\">⊖</a><span class=\\"object syntax\\">{</span><span>
</span><span>        </span><span class=\\"key\\">\\"title\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"example glossary\\"</span><span class=\\"syntax\\">,</span>
<span>        </span><span class=\\"key\\">\\"comprehensive\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"boolean\\">true</span><span class=\\"syntax\\">,</span>
<span>        </span><span class=\\"key\\">\\"link\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"keyword\\">undefined</span><span class=\\"syntax\\">,</span>
<span>        </span><span class=\\"key\\">\\"count\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"number\\">1</span><span class=\\"syntax\\">,</span>
<span>        </span><span class=\\"key\\">\\"GlossDiv\\"</span><span class=\\"object syntax\\">: </span><span><span class=\\"object\\" style=\\"display: none;\\"><a class=\\"disclosure\\" href=\\"#\\">⊕</a><span class=\\"object syntax\\">{</span><a href=\\"#\\">2 items</a><span class=\\"object syntax\\">}</span></span><span class=\\"object\\" style=\\"display: inline;\\"><a class=\\"disclosure\\" href=\\"#\\">⊖</a><span class=\\"object syntax\\">{</span><span>
</span><span>            </span><span class=\\"key\\">\\"title\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"S\\"</span><span class=\\"syntax\\">,</span>
<span>            </span><span class=\\"key\\">\\"GlossList\\"</span><span class=\\"object syntax\\">: </span><span><span class=\\"object\\" style=\\"display: none;\\"><a class=\\"disclosure\\" href=\\"#\\">⊕</a><span class=\\"object syntax\\">{</span><a href=\\"#\\">1 item</a><span class=\\"object syntax\\">}</span></span><span class=\\"object\\" style=\\"display: inline;\\"><a class=\\"disclosure\\" href=\\"#\\">⊖</a><span class=\\"object syntax\\">{</span><span>
</span><span>                </span><span class=\\"key\\">\\"GlossEntry\\"</span><span class=\\"object syntax\\">: </span><span><span class=\\"object\\" style=\\"display: none;\\"><a class=\\"disclosure\\" href=\\"#\\">⊕</a><span class=\\"object syntax\\">{</span><a href=\\"#\\">7 items</a><span class=\\"object syntax\\">}</span></span><span class=\\"object\\" style=\\"display: inline;\\"><a class=\\"disclosure\\" href=\\"#\\">⊖</a><span class=\\"object syntax\\">{</span><span>
</span><span>                    </span><span class=\\"key\\">\\"ID\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"SGML\\"</span><span class=\\"syntax\\">,</span>
<span>                    </span><span class=\\"key\\">\\"SortAs\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"SGML\\"</span><span class=\\"syntax\\">,</span>
<span>                    </span><span class=\\"key\\">\\"GlossTerm\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"Standard Generalized Markup Language\\"</span><span class=\\"syntax\\">,</span>
<span>                    </span><span class=\\"key\\">\\"Acronym\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"SGML\\"</span><span class=\\"syntax\\">,</span>
<span>                    </span><span class=\\"key\\">\\"Abbrev\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"ISO 8879:1986\\"</span><span class=\\"syntax\\">,</span>
<span>                    </span><span class=\\"key\\">\\"GlossDef\\"</span><span class=\\"object syntax\\">: </span><span><span class=\\"object\\" style=\\"display: none;\\"><a class=\\"disclosure\\" href=\\"#\\">⊕</a><span class=\\"object syntax\\">{</span><a href=\\"#\\">2 items</a><span class=\\"object syntax\\">}</span></span><span class=\\"object\\" style=\\"display: inline;\\"><a class=\\"disclosure\\" href=\\"#\\">⊖</a><span class=\\"object syntax\\">{</span><span>
</span><span>                        </span><span class=\\"key\\">\\"para\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"A meta-markup language, used to create markup languages such as DocBook.\\"</span><span class=\\"syntax\\">,</span>
<span>                        </span><span class=\\"key\\">\\"GlossSeeAlso\\"</span><span class=\\"object syntax\\">: </span><span><span class=\\"array\\" style=\\"display: none;\\"><a class=\\"disclosure\\" href=\\"#\\">⊕</a><span class=\\"array syntax\\">[</span><a href=\\"#\\">2 items</a><span class=\\"array syntax\\">]</span></span><span class=\\"array\\" style=\\"display: inline;\\"><a class=\\"disclosure\\" href=\\"#\\">⊖</a><span class=\\"array syntax\\">[</span><span>
</span><span>                            </span><span class=\\"string\\">\\"GML\\"</span><span class=\\"syntax\\">,</span>
<span>                            </span><span class=\\"string\\">\\"XML\\"</span>
<span>                        </span><span class=\\"array syntax\\">]</span></span></span>
<span>                    </span><span class=\\"object syntax\\">}</span></span></span><span class=\\"syntax\\">,</span>
<span>                    </span><span class=\\"key\\">\\"GlossSee\\"</span><span class=\\"object syntax\\">: </span><span></span><span class=\\"string\\">\\"markup\\"</span>
<span>                </span><span class=\\"object syntax\\">}</span></span></span>
<span>            </span><span class=\\"object syntax\\">}</span></span></span>
<span>        </span><span class=\\"object syntax\\">}</span></span></span>
<span>    </span><span class=\\"object syntax\\">}</span></span></span>
<span></span><span class=\\"object syntax\\">}</span></span></span>"
`;
